### ISSUE TITLE
FE: Discovery: do all popular mobile browsers have a way to display app store links like iOS Safari does?

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#15339]
### User story
> _Describe the audience/user, enhancement or fix, and value / outcome desired._

**AS A**
**I WANT**
**SO THAT**

## Acceptance criteria

>Consider
> - Design / Accessibility reviews
> - Collab cycle requirements
> - Device sizes (mobile first)
> - Documentation updates / Change management
< - [Content model documentation](https://prod.cms.va.gov/admin/structure/cm_document)
> - Testing notes

- [ ]

**Comment by wesrowe** [comment]
@chriskim2311, please pre-refine this ticket for reasonableness, any other input you have.

**Comment by chriskim2311** [comment]
@wesrowe I know that safari is the only one that has a built in option for iOS devices. Android and other browsers don't have a "easy" way to implement like Safari.

I did find this npm package we could use https://www.npmjs.com/package/smartbanner.js?amp=1 
It takes care of all browsers other than Safari and tackles android as well. 

It does seem maintained, but what are VA thoughts on using third-party packages for implementations like this?

**Comment by wesrowe** [comment]
Thanks for the quick research @chriskim2311. I think the npm question is one for Platform. Is it possible the answer is in Platform docs?

**Comment by jilladams** [comment]
From FE refinement: 
* Medium article is required reading before beginning work. 
* Medium article solution always took users to the app store, regardless of whether app was installed. 
  * We will need to decide what the requirement is for that (ok to app store or no, should go to the app if it's installed).
  * We don't want to send Veterans to the app store if they already have the app. We think this may be the default behavior 
* Randi noted there are other NPM options, etc., that may solve for this, besides what's in the article.
* Accessibility Q: move from webpage to an app: is that UX a problem? This is not new to this initiative, so prob outside our scope to be concerned with. Is ultimately a mobile app team UX issue.

**Comment by chriskim2311** [comment]
For this ticket I investigated and implemented Smartbanner.js.

I believe this has the most built in experience and I was able to solve the issues with the a11y concerns. I have screenshots of the banner for Android and iOS devices. 

To get this to work I have implemented the following:
- Using unpkg.com I added a link and script tag to the `header.html` file in content-build. These tags will load the assets needed to run the package and the styles that go with it.
```
  <link rel="stylesheet" href="https://unpkg.com/smartbanner.js/dist/smartbanner.min.css">
  <script src="https://unpkg.com/smartbanner.js/dist/smartbanner.min.js"></script>  
```
- Using this NPM package we need to set the meta tags that the package will pull from. I have added these tags into the `metatags.drupal.liquid` file. Here we are previously using logic to render the safari banner to the specific URLs in a previous ticket. We will utilize the same logic to render the meta tags only for these URLs. 
```
  <!-- Start SmartBanner configuration -->
  <meta name="smartbanner:exclude-user-agent-regex" content="^.*(Version).*Safari">
  <meta name="smartbanner:title" content="VA: Health and Benefits">
  <meta name="smartbanner:author" content="US Dept. of Veteran Affairs">
  <meta name="smartbanner:price" content=" ">
  <meta name="smartbanner:price-suffix-apple" content=" ">
  <meta name="smartbanner:price-suffix-google" content=" ">
  <meta name="smartbanner:icon-apple" content="https://is1-ssl.mzstatic.com/image/thumb/Purple116/v4/cb/43/1f/cb431fa1-b110-805f-251d-2fbb68babb1a/AppIcon-1x_U007emarketing-0-7-0-85-220.png/230x0w.webp">
  <meta name="smartbanner:icon-google" content="https://play-lh.googleusercontent.com/pxpOEzV8odLxElJtNrPhorM48TPh6p-RnTWqwvlhetuMM3lZuGmbXP2_4YHlQ-fCOA=w240-h480-rw">
  <meta name="smartbanner:button" content="VIEW">
  <meta name="smartbanner:button-url-apple" content="https://apps.apple.com/us/app/va-health-and-benefits/id1559609596">
  <meta name="smartbanner:button-url-google" content="https://play.google.com/store/apps/details?id=gov.va.mobileapp&hl=en_US&gl=US&pli=1">
  <meta name="smartbanner:enabled-platforms" content="android,ios">
  <meta name="smartbanner:close-label" content="Close">
  <!-- End SmartBanner configuration -->
```
- To handle the a11y issue I have implemented a secondary function that will move the banner div into the appropriate order in the DOM.
```
  <script  type="text/javascript" defer>
    window.onload = function() {
      const bodyDiv =  document.querySelector('body');
      const contentElement = document.querySelector('.smartbanner');
      if(contentElement){
      contentElement.style.top = '-84px'
       bodyDiv.prepend(contentElement)
      }
    };
  </script>
```

**iOS Chrome Browser**
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOEpXQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--369cb2405b4c1916ac03bd79576ceb274621f486/iPhone_13_v16_3.png)

**Android Chrome Browser**
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOGxXQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--ed866056067aac44e0afadc519a22191000c320f/Samsung_Galaxy_S22_Ultra_v12_0_and_Kevin_Porter_Jr__s_ex-girlfriend_refutes__false_narrative__that_he_beat__strangled_her.png)

**Android Firefox Browser**
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBODFXQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8b64e6671132b3d0e2602499b7b354d5e500984f/Samsung_Galaxy_S22_Ultra_v12_0.png)

**Comment by chriskim2311** [comment]
@thejordanwood  @jilladams @FranECross @wesrowe 
Hi all! Please take a look at the screenshots and I can make changes as necessary. Thanks!

@laflannery: I can provide a ngrok link if you would like to take a look at the tabbing issue. 
The smartbanner is now the first DOM element in the body.
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOWhXQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f62734f86a1300fdb08d7ec3cec7aee538bb9397/VA_Prescription_Refill_And_Tracking___Veterans_Affairs.png)

**Comment by laflannery** [comment]
@chriskim2311 The DOM looks great, The NGROK link would be good just to confirm if it's not too much work

**Comment by chriskim2311** [comment]
@thejordanwood Good question! The android banner is similar to what an android user would see for their app store, but not  sure if we want to show the Apple store banner style to an android user?

**Comment by jilladams** [comment]
@chriskim2311 just flagging Laura's note above re: an NGROK link in case you didn't already send that elsewhere. Also: when the step for checking re: 3rd party lib inclusion happens with Platform, could you link it here for future ref?

**Comment by chriskim2311** [comment]
@jilladams Yes I was able to get the ngrok over to Laura! And I did create the platform support ticket waiting on their review.

https://dsva.slack.com/archives/CBU0KDSB1/p1697826712847279

**Comment by rmessina1010** [comment]
My only concern , other than the tab order,  which you seem to have solved would be styling differences on iOS Chrome, is this a  was this just not on "dark mode?" or are styling failing to apply?

**Comment by FranECross** [comment]
Dave Conlon messaged in [Slack](https://dsva.slack.com/archives/C52CL1PKQ/p1698164949100899) asking if we can use Android styling for the Android app

**Comment by FranECross** [comment]
@chriskim2311 Amanda and Dave have both given their approval to use the default Android styling for the mobile banner. Thanks! Here's the [Slack ](https://dsva.slack.com/archives/C52CL1PKQ/p1698167900654979?thread_ts=1698164949.100899&cid=C52CL1PKQ)thread with their approvals. cc @jilladams

**Comment by jilladams** [comment]
Platform still evaluating any steps needed for signing off on 3rd party JS addition to the code. This will stay open til that lands. (Same Plat thread: https://dsva.slack.com/archives/CBU0KDSB1/p1697826712847279)

**Comment by jilladams** [comment]
SmartBanner.js approved by Platform! https://dsva.slack.com/archives/CBU0KDSB1/p1699377786208839?thread_ts=1697826712.847279&cid=CBU0KDSB1. Closing.

FYI @davidconlon @FranECross @chriskim2311
